### PR TITLE
NETBEANS-1499 - change package.json icon

### PR DESCRIPTION
**Pull request by Chris2011**
This PR will change the generic js icon for a specific file (package.json) with a specific mimetype (text/package+x-json) to the logo of npm, to make it more prominent.

Before merging, we have to check the licenses of the npm logo: https://www.npmjs.com/policies/trademark. I wrote them an e-mail, will let you know whether it is ok or not. Maybe a license guy can help here too.

This logo is also used inside of a VS Code plugin for better icons for specific filetypes.

### CODE REVIEW

**Comment by emilianbold** [comment]
Seems to me the PR does more than just change an icon.

**Comment by Chris2011** [comment]
@emilianbold yes, I had to create a DataObject for the file to set a different icon. Otherwise it wont work. Or maybe I didn't know any other solution, maybe there is another one :). But it seems that I can't use it, I'm in contact with the tech support for npm. Will let you know.

**Comment by Chris2011** [comment]
Today I got the permission to use the npm logo:

> Hi, Chris!  Sorry for the delay.  I wanted to loop our trademark and copyright counsel in on this, just to make sure I wasn't doing something dumb :)

>The Apache Netbeans IDE has asked for permission to use the npm logo to represent the filetype of files named "package.json" and "package-lock.json" in the IDE. Thanks for asking, and yes, you have our permission to do so. You also linked the logo at https://github.com/vscode-icons/vscode-icons/blob/master/icons/file_type_npm.svg, which is the icon that you may use.
Thanks for your patience and for reaching out to us about this matter.  Looking forward to seeing the npm logo in Netbeans :)

>Cheers!

So pleaes have a look and leave comments.

**Comment by Chris2011** [comment]
Summerize:
* Created a new DataObject for "package.json" called PackageJsonDataObject to add a new file icon.
* Removed old layer.xml file
* Moved annotation for MimeResolver from PackageJson.java to PackageJsonDataObject.java
* Added the npm icon to the code. I got the permission to use it.

Let me know, if anything is still unclear.

**Comment by Chris2011** [comment]
Here you can see the result:

image

**Comment by BigChief45** [comment]
I am not very familiar with JavaScript package managers, but isn't package.json also used by other package managers such as Yarn?

**Comment by Chris2011** [comment]
Yes, and for pnpm too, but initially it came from npm. yarn also uses node_modules, pnpm too. So from the core they are pretty much the same. The difference is how they treat the modules. Hard linking, caching, etc. So in general if someone will see the npm logo or package.json, if they google for it, they will figure it out quickly. https://www.google.com/search?q=package.json

**Comment by matthiasblaesing** [comment]
From a technical perspective I think this is good to go. I ran the build and found, that the new icon is present in the package.json file, the file still has syntax highlighting and the node actions are also correctly present.

We need to be careful with the artwork though. We already need to discuss enough about licenses. For external files we record the information in a file named licenseinfo.xml - see details here:
https://cwiki.apache.org/confluence/display/NETBEANS/Legal+FAQ#LegalFAQ-licenseinfoPerfileinfo-licenseinfo.xml
There are several samples in the codebase for the file. Having the file ensures, that the license for the graphics is correctly included in the summary and if needed attributions are included in the NOTICE file.

**Comment by Chris2011** [comment]
@matthiasblaesing or others, so what is needed here to merge the PR or change it? Would like to add it to the next NB version.

**Comment by matthiasblaesing** [comment]
@Chris2011 did you see my previous comment?
1. Add/Modify licenseinfo.xml to reflect the license info for the graphics file you are adding (this is covered in my previous comment with links for infos)
2. Squash the commit into one (can be done on merge of course, I personally prefer, that I have control over the commits as the author)

**Comment by Chris2011** [comment]
@matthiasblaesing ok thix, thought that this was only a discussion topic and for me an info. Will change it. Thx.

Should I squash all commits into one? Or do you mean only the license changing?

**Comment by matthiasblaesing** [comment]
@Chris2011 from my POV the commits form one logical change and thus should all squashed together.

**Comment by Chris2011** [comment]
Ok, thx. That makes sense.

**Comment by Chris2011** [comment]
Am I right, if I say that the license of the logo is creative common? https://www.npmjs.com/policies/trademark I can't find any other info about that. @matthiasblaesing what do you think? I know that license for NPM is "Artistic License 2.0" Is this the same for the logo?

**Comment by Chris2011** [comment]
This will be the code, that I want to add it into the licenseinfo.xml file inside of the Node.js Project Support module
```xml
    <fileset>
        <file>src/org/netbeans/modules/javascript/nodejs/resources/npm.png</file>
        <license ref="Artistic License 2.0" />
        <comment type="COMMENT_UNSUPPORTED" />
    </fileset>
```

**Comment by matthiasblaesing** [comment]
@Chris2011 the license for the logo file might not be the same, so you need to check: Where did you get the graphics and which license is referenced there? This might be implicit (from the context of the project) or explicit from a concrete information.

The license ref entry must match an entry in the nbbuild/licenses directory, else it will be rejected.

**Comment by Chris2011** [comment]
Thx, will check.

**Comment by isaacs** [comment]
I'm a little bit confused by the questions here.  The logo is not "software", it's an image, and so there's a bit of a category error talking about what OSI license it uses.  The logo is not covered by Artistic 2.0.  IANAL, but I believe that even if it were covered under that license, it would cover the specific bits distributed (which you aren't using), not the design as such (which is a mark used in trade to refer to a specific entity).

It's governed by trademark, and our trademark policy says in a nutshell, "all rights reserved", and thus may not be used without permission for anything other than nominative use or fair use (which are protected by law anyway).  It's certainly not Creative Commons.

One could make the case that this is nominative use.  But you don't have to, because we've provided written permission.  I don't know how it can be more clear that you're allowed to use it than the trademark owner saying in writing that they give you permission, but maybe I'm missing something.

**Comment by matthiasblaesing** [comment]
Of course the logo is covered under a license. I can only speak for germany, if I don't give you a license to use my work, than can't use it. I don't need to do anything to protect it. I'll assume, that it is similar in other countries. That is independed of trademark law, which from my POV in this case does not even apply, as the icon only symbolizes the application the file belongs to.

I'll make a sample. Consider font-awesome:

https://github.com/FortAwesome/Font-Awesome/blob/master/svgs/brands/npm.svg

The file is covered by the CC-BY-4.0 license:

https://github.com/FortAwesome/Font-Awesome#license

So we could render that, add attribution for it to the apache notice file and be done with it.

For a list of acceptable licenses, please see the apache legal faq:
Good: https://www.apache.org/legal/resolved.html#category-a
Problematic, but possible: https://www.apache.org/legal/resolved.html#category-b

**Comment by isaacs** [comment]
The license is "All rights reserved".  But we gave you (that is, Apache NetBeans) written permission to use the logo in the way described.

If you want to put that in some more agreeable legal language for us to sign, go for it.  I don't want to seem rude here, but as this is a favor to Netbeans rather than the other way around, I'm not super interested in investing much energy in making this happen.

**Comment by matthiasblaesing** [comment]
@Chris2011 please find a logo file with a clearly stated license. A license "for us" is not enough. It has to be compatible with the ASFs requirements for licenses (I inlined the white-listed licenses above) - I don't want to involve apache legal, just to get one icon included. I would prefer a source, that does not require attribution and licenses the logo according to category-a above.

**Comment by emilianbold** [comment]
> A license "for us" is not enough. It has to be compatible with the ASFs requirements for licenses (I inlined the white-listed licenses above)

Good point. An Apache Licensed projects allows all kinds of usage downstream, including commercial. So, we can't have a logo that only "NetBeans" can use, because then we basically cannot release that.

From reading https://www.npmjs.com/policies/trademark 

> The npm Logo signifies us, or a special relationship with us, and you should use it only with our permission. Since the goal is to avoid confusion about you being us, or your relationship with us, context counts. We will consider requests on a case-by-case basis.

it seems clear that we cannot include the logo in NetBeans as each NetBeans derivative will have to get specific permission for the npm logo.

Since I don't believe npm inc will grant us such a generous permission given the terms in their trademark policy we can probably reject this PR.

I'm also thinking if we really need the npm logo there. We could differentiate the file from a plain .js file with some hint. Even for Maven projects, we don't have a specific icon for pom.xml, we have the normal XML icon and we just group the pom.xml file under the "Project Files" subtree.

**Comment by isaacs** [comment]
> Since I don't believe npm inc will grant us such a generous permission given the terms in their trademark policy

Indeed, that would not be something we would be comfortable with.  Some downstream uses could imply a relationship with npm, Inc. or the npm project, and would thus cause confusion in the marketplace.

**Comment by Chris2011** [comment]
@isaacs I'm really sry, that this is such annoying, for me too. It seems everything is about license and at the end, nothing is about license. But thx for you help anyway.

To let you know, the icon is the offficial one, how should I find another one? It will be a custom one, from someone else, right? Or should I create my own based on that? If this is legally possible, I can do.

> I'm also thinking if we really need the npm logo there. We could differentiate the file from a plain .js file with some hint. Even for Maven projects, we don't have a specific icon for pom.xml, we have the normal XML icon and we just group the pom.xml file under the "Project Files" subtree.

Of course @emilianbold we have, beacuse I implemented it. It is a visual thing: https://github.com/apache/incubator-netbeans/pull/983 and there are lots of more coming: https://issues.apache.org/jira/browse/NETBEANS-1496

Putting under the project files is not enough in my POV, as I said this is a visual thing, if you see the name, maybe you heard about it, but if you see the logo, you can guess that this is related to npm and for pom, it is related to maven etc. So yes I want to change some icons, which are only normal JS icons (also for JSON files, which doesn make sense but this is an other topic) to make them more promiment. I couldn't know that this is such a terrible process to find the right license or whatever.

NetBeans still uses the karma logo for the test runner and where did you guys get it? What license is that? I

**Comment by matthiasblaesing** [comment]
To make this easier - this looks like a good alternative:

https://github.com/simple-icons/simple-icons/blob/develop/icons/npm.svg

That file is CC-0 (see: https://github.com/simple-icons/simple-icons/blob/develop/LICENSE.md). I assume, that black is not the color you want. So I would do this:

1. Get the file referenced above
2. Modify it with an SVG Editor and change color to your liking
3. Add a XML comment with the ALv2 (ASF donation variant) license before the opening svg-tag
4. Add a second comment, that states (preferably as bullet points / numbered list): where did you download the original document (URL above), that the original file was CC-0 licenses, that you modified color and added license comment
5. Render the necessary icon file from it, please the icon file and the svg file together in the resources package (for the package I suggest org.netbeans.modules.javascript.nodejs.ui.resources (https://github.com/apache/incubator-netbeans/tree/b48eb9b1cc267f7e7c768f6e90c4d539ee47d19d/webcommon/javascript.nodejs/src/org/netbeans/modules/javascript/nodejs/ui/resources), as there are already icons
6. Add an entry into the licenseinfo.xml (https://github.com/apache/incubator-netbeans/blob/b48eb9b1cc267f7e7c768f6e90c4d539ee47d19d/webcommon/javascript.nodejs/licenseinfo.xml). As the file is now ALv2, you can just add another file entry to the existing fileset for the png/gif (the svg can hold a license)

This way we can skip the whole license discussion and get on.

**Comment by Chris2011** [comment]
@matthiasblaesing ok thx. Will do so. Thx for the research.

**Comment by emilianbold** [comment]
npm Inc still owns the trademark to that logo, regardless who designed an icon representing their logo and under what license the icon itself is. So... this doesn't really change the trademark policy angle.

Also, package.js having the npm logo is not in line with how NetBeans displayed such files before. The Ant / Maven configuration files don't have specific icons! It's all in the tree structure and in how the projects are displayed ("Project Files", etc).

So considering the npm inc trademark angle *and* this not being something NetBeans does, I would think of working at project level to group / present the package.js file in a different subtree, as a JS file (perhaps with a small decoration in some icon corner).

Needless to say package.js is not "owned" by npm. It's quite possible to have projects that have a package.js without npm (yarn supposedly reads it and any other tool that wants to work with existing projects will have to support package.js).

**Comment by matthiasblaesing** [comment]
I don't see trademark as not relevant here. Everything that was said regarding trademark here is also true for Google for Chrome, Microsoft for IE, Mozilla for Firefox, Oracle for Java and so on.
Apache legal requires us to comply with the licenses and that is the case for the file I mentioned in my previous comment. Independend from that trademark laws may or may not apply. For our use case (icon for a file that is primarily used by the programm that uses that logo), I consider it nominative use, which you can't prevent as the trademark holder.

If we really want to discuss the pro/con of using the NPM logo to represent package.json, I suggest to shift discussion to the mailinglist. My review assumes, that we aggree, that using the logo is a good idea to make the file easily recognizable.

**Comment by emilianbold** [comment]
OK, no need for me to get in the way. Carry on...

**Comment by isaacs** [comment]
Why not just use the yellow-square JS logo? It’s explicitly public domain,
I believe, and saying “this is for JS” is probably about as meaningful for
end users as saying it’s for npm.

**Comment by isaacs** [comment]
And, just on behalf of the USA and capitalism, I apologize that trademark
and copyright can be such a pita. It does seem like the right thing would
be to allow this use, which is why we gave permission, but we can’t make it
too broad without losing our ability to control who can claim to speak for
us.

I understand the concern from the Apache side, and I appreciate everyone
trying to do the right thing here.

**Comment by Chris2011** [comment]
> Why not just use the yellow-square JS logo? It’s explicitly public domain, I believe, and saying “this is for JS” is probably about as meaningful for end users as saying it’s for npm.

I don't know whether it was a bit of sarcastic or not, if it was sarcastic, don't read further :), but we already have icons for JS and unfortunately the same for json files, which I also will change, but for this I will make my own with my UX/UI friend David.

Thx for everyone, if everything is fine now, I will use the approach of @matthiasblaesing or should we discuss it further?

**Comment by isaacs** [comment]
Oh! No it wasn’t sarcastic. Sorry, I didn’t know that you already had an
icon for JS. Can you use the js icon for package.json as well, then?

**Comment by Chris2011** [comment]
> Oh! No it wasn’t sarcastic. Sorry, I didn’t know that you already had an icon for JS. Can you use the js icon for package.json as well, then?

We already have that, we have Node and NPM and Bower support and all JS and all JSON files, have the same JS icon. This is what I wanted to change because some files, like bower.json, gruntfile.js, package.json, webpack.config.js, karma.config.js etc. should be more prominent. As this is doing the same for VS Code: https://github.com/vscode-icons/vscode-icons

Please see the gif how it will look the package.json there. First IMHO that JS and JSON files have the same icon is not correct, but this is not going to work here. Second and this is my solution, I want to change some icons for specific files for some technologies. I mean sure we can made all files, doesn't matter which mimetype/filetype make the same icon, blank white icon (this was sarcastic).

**Comment by Chris2011** [comment]
To be clear, I don't want it to make it the same as VS Code, I only see the benefit out of it and I change the icon for the pom.xml file too as IntelliJ has another icon for that specific file too. I didn't use the same icon for the pom.xml because I already have one and I used that. But for package.json, bower.json and the others, we don't have different icons. Different projects have different icons, different files have different icons, etc.

**Comment by Chris2011** [comment]
image

This is how it looks now. Each JS and JSON file have the same JS icon and with my approach the package.json is different, which you can catch with your eye, very well that this file is more important or different from each other. And it has a different mimetype, due to special actions on that file.

**Comment by Chris2011** [comment]
Ok I understand, so we have to clarify here, that we are talking about NetBeans IDE. But there is a NetBeans Platform where you can create applications, based on NetBeans. I hope that the icon will not part of the platform, or? @matthiasblaesing. Then we have a problem or would a permission to the icon using in the netbeans platform is ok?

**Comment by matthiasblaesing** [comment]
Ok, this was raised with apache legal:
https://issues.apache.org/jira/browse/LEGAL-353
I seriously hope, that we are prepared for the answer....

**Comment by matthiasblaesing** [comment]
Ok - Apache legal gave its verdict: As long as we use the icon in the way we plan (as an icon representing the programm), it is fair use. We also need a valid license, that satisfies the requirements of the apache foundation regarding the copy right. Using the CC-0 licensed base will keep the copy right easy and the trademark part is covered by fair use. @Chris2011 is suggest, you follow the advise from https://github.com/apache/incubator-netbeans/pull/1034#issuecomment-453753472

**Comment by Chris2011** [comment]
hey @matthiasblaesing great news. Thx. Will do so. But i want to add the icon to the package-lock.json and npm-debug.log. Do I need to ask again for permission for those files? I mean those files are also part of that and shouldn't be a problem, right?

**Comment by matthiasblaesing** [comment]
I would be ok with using the icon for the two files. There is a direct relationship from `package.json`, `package-lock.json` and `npm-debug.log` to npm. Using the icon seems logical.

**Comment by isaacs** [comment]
For what it's worth (I guess not much if it's fair use anyway), I think it's completely reasonable to use the same logo for package.json as package-lock.json and npm-debug.log files.

Glad to hear this reached a reasonable resolution :)

**Comment by Chris2011** [comment]
Great thx to @isaacs for your help and patience and thx to @matthiasblaesing for the help and the legal part. Will do it in the next couple of days :)

**Comment by matthiasblaesing** [comment]
@Chris2011 to me the change looks sane. Could you rebase the changeset onto master and squash the changeset? That way we get one or two meaningful commits and it makes it easier to cherry-pick the change, if needed (see here: https://lists.apache.org/thread.html/c9087dec90200badd2528dfd1c315d2c619ee3dadcf735989914fb84@%3Cdev.netbeans.apache.org%3E). I would reword it a bit and say: Make meaningful self contained commits.

**Comment by Chris2011** [comment]
Hey @matthiasblaesing I'm not done yet, unfortunately. I have a big problem that needs to figure out first. The source/history tab pane is missing, after I registered the PackageJsonDataObject:

image


After I fix this, I will add the icon for the npm-debug.log and the package-lock.json. When I'm finish, I will squash my commits and see how it will work. Thx for the comment :)

**Comment by Chris2011** [comment]
Finally. I did it. Everything ok now?

File: webcommon/javascript.nodejs/src/org/netbeans/modules/javascript/nodejs/resources/npm-resolver.xml [hunk 1]
```diff
@@ -26,4 +26,12 @@
         <name name="package.json"/>
         <resolver mime="text/package+x-json"/>
     </file>
+    <file>
+        <name name="package-lock.json"/>
+        <resolver mime="text/package-lock+x-json"/>
+    </file>
+    <file>
+        <name name="npm-debug.log"/>
+        <resolver mime="text/plain"/>
```
Comment (matthiasblaesing) [inline webcommon/javascript.nodejs/src/org/netbeans/modules/javascript/nodejs/resources/npm-resolver.xml]: This looks fishy. I understand, that you want to associate your new Editor with only "npm-debug.log", yet you assign it the mimeType "text/plain". That mimetype matches a big chunk of files on my harddisk, which you would all pull through your special data object.

Comment (Chris2011) [reply webcommon/javascript.nodejs/src/org/netbeans/modules/javascript/nodejs/resources/npm-resolver.xml]: Oops, I didn't know that. I thought the combination of resolver AND filename in this example is enough to make it unique. So for this I really need a new mimetype, right? text/package+plain or text/npm-debug-log. Looks weirdo. Any other solutions?

Comment (matthiasblaesing) [reply webcommon/javascript.nodejs/src/org/netbeans/modules/javascript/nodejs/resources/npm-resolver.xml]: None I'm aware of - this file just describes how a mimetype is assigned. In your case you are using a name to mime-type mapping.

Comment (Chris2011) [reply webcommon/javascript.nodejs/src/org/netbeans/modules/javascript/nodejs/resources/npm-resolver.xml]: Ok, will change the mimetype, weird but ok. I think it must be possible to add more stuff to a fully qualified unique file name like package.json or in my example to npm-debug.log like adding icons to not change or adding a mimetype. But this is another topic. I will change the mimetype to text/npm-log if this is okish.

**Comment by Chris2011** [comment]
I changed it, but I thought I did it correct, now I force pushed it, sry. I fixed it now.

**Comment by matthiasblaesing** [comment]
Looks good to me.

**Comment by Chris2011** [comment]
Rebased as there are no other remarks.

**Comment by matthiasblaesing** [comment]
I looked at the mime registration again today and while reading about this, I noticed, that the mimetypes should have been different:

- `text/package+x-json` should have been `application/x.npm-package+json`
- `text/package-lock+x-json` should have been `application/x.npm-package-lock+json`
- `text/npm-log` should have been `text/x.npm-log+plain`

Reasoning: The mimetypes are unregistered (thus the `x.` for the subtype) but they are all based on existing types. For former two are based on `application/json` and the latter is based on `text/plain`. The `+json` parts are the suffix to designate their base type.

Its to late now, but should be considered next time.
